genbindings.sh

```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env bash
+
+# Generate (and reasonably test) C bindings
+
+set -e
+# First build the latest c-bindings-gen binary
+cd c-bindings-gen && cargo build && cd ..
+
+# Then wipe all the existing C bindings (because we're being run in the right directory)
+rm -rf lightning-c-bindings/src/{ln,util,chain,routing}
+mkdir -p lightning-c-bindings/src/{ln,util,chain,routing}
+
+# Finally, run the c-bindings-gen binary, building fresh bindings.
+SRC="$(pwd)/lightning/src"
+OUT="$(pwd)/lightning-c-bindings/src"
+OUT_TEMPL="$(pwd)/lightning-c-bindings/src/c_types/derived.rs"
+OUT_F="$(pwd)/lightning-c-bindings/include/rust_types.h"
+OUT_CPP="$(pwd)/lightning-c-bindings/include/lightningpp.hpp"
+RUST_BACKTRACE=1 ./c-bindings-gen/target/debug/c-bindings-gen $SRC/ $OUT/ lightning $OUT_TEMPL $OUT_F $OUT_CPP
+
+# Now cd to lightning-c-bindings, build the generated bindings, and call cbindgen to build a C header file
+PATH="$PATH:~/.cargo/bin"
+cd lightning-c-bindings
+cargo build
+cbindgen -v --config cbindgen.toml -o include/lightning.h
+
+# cbindgen is relatively braindead when exporting typedefs -
+# it happily exports all our typedefs for private types, even with the
+# generics we specified in C mode! So we drop all those types manually here.
+sed -i 's/typedef LDKln.*Import.*LDKln.*;//g' include/lightning.h
+
+# Finally, sanity-check the generated C and C++ bindings with demo apps:
+
+# Naively run the C demo app:
+gcc -Wall -g -pthread demo.c ../target/debug/liblightning.a -ldl
+./a.out
+
+# And run the C++ demo app in valgrind to test memory model correctness
+# Note that there are a few leaks currently, so we don't fail the script for leaks,
+# but ideally we shouldn't add new leaks and once we fix the existing ones we should
+# fail in response to any memory leaks.
+g++ -Wall -g -pthread demo.cpp -L../target/debug/ -llightning -ldl
+LD_LIBRARY_PATH=../target/debug/ valgrind --error-exitcode=4 --memcheck:leak-check=full --show-leak-kinds=all ./a.out
+
+# Test a statically-linked C++ version, tracking the resulting binary size and runtime 
+# across debug, LTO, and cross-language LTO builds (using the same compiler each time).
+clang++ -Wall -pthread demo.cpp ../target/debug/liblightning.a -ldl
+./a.out
+echo " C++ Bin size and runtime w/o optimization:"
+ls -lha a.out
+time ./a.out > /dev/null
+
+# Now build with LTO on on both C++ and rust, but without cross-language LTO:
+cargo rustc -v --release -- -C lto
+clang++ -Wall -flto -O2 -pthread demo.cpp ../target/release/liblightning.a -ldl
+echo "C++ Bin size and runtime with only RL (LTO) optimized:"
+ls -lha a.out
+time ./a.out > /dev/null
+
+# Finally, test cross-language LTO. Note that this will fail if rustc and clang++
+# build against different versions of LLVM (eg when rustc is installed via rustup
+# or Ubuntu packages). This should work fine on Distros which do more involved
+# packaging than simply shipping the rustup binaries (eg Debian should Just Work
+# here).
```

Comment:
I'm able to build even when `rustc` is installed with `rustup` (though it could be because coincidentally my system has a compatible version of LLVM). Any more instructions to reproduce (version of ubuntu, etc)?

Also, asked Carl about this issue (cuz build systems) and he suggested that it'd be worth re-testing this problem with the flag mentioned in my other review comment, to see if perhaps that's the problem (which makes sense to my limited understanding)?

Reply:
Ubuntu with rustup blows up with `/usr/bin/ld: error: Failed to link module ../target/release/liblightning.a.llvm.7177896.compiler_builtins.ddmutrpc-cgu.92: Invalid summary version 8, 1, 2, 3 or 4 expected` on Github Actions.

Reply:
FWIW, I haven't been able to successfully run `genbindings.sh` on MacOS without significant modifications.

Reply:
Is it specifically because of this? We could make the cross-language LTO step an optional step since its really just another Cool Thing, not really a critical test.

Reply:
I wonder whether simply given the presence of genbindings.sh, the lightning-c-bindings directory should be tracked by git?